overlord/state/change.go

```diff
@@ -422,13 +423,35 @@ func (c *Change) Status() Status {
 	panic(fmt.Sprintf("internal error: cannot process change status: %v", statusStats))
 }
 
+func (c *Change) notifyStatusChange(new Status) {
+	if c.lastObservedStatus == new {
+		return
+	}
+	c.state.notifyChangeStatusChangedObservers(c, c.lastObservedStatus, new)
+	c.lastObservedStatus = new
+}
+
 // SetStatus sets the change status, overriding the default behavior (see Status method).
 func (c *Change) SetStatus(s Status) {
 	c.state.writing()
-	c.status = s
-	if s.Ready() {
-		c.markReady()
+
+	if s != DefaultStatus {
+		if s.Ready() {
+			c.markReady()
+		}
+		c.notifyStatusChange(s)
+		c.status = s
+		return
 	}
+
+	// Set new status immediately to calculate the
+	// aggregated status if s == DefaultStatus.
+	// We don't need to worry about markReady here,
+	// because if the change became ready under the
+	// fixed status, markReady will still have been called
+	// by taskStatusChanged once all the tasks are ready.
```

Comment:
this comment is a bit confusing, because in the end, comparing with the old code we don't need to call markReady because DefaultStatus.Ready() is false, no? anyway I think we could make the diff to this function smaller, as I'm not sure why we want to avoid calling s.Ready and make a difference between DefaultStatus or not. Just adding notify to the old code would seem ok?

Reply:
I changed this and simply added the notify. I was trying to cover a case that doesn't exist, which probably didn't make sense - I instead added a unit test to cover the how it works today by using the new "observer" code in state_test.go